File: ch2/lineards/list.py
from collections import deque 

def main():
    s = []
    print("{:d}".format(len(s) == 0))            # s is empty, true (1)
    print("==================")
    s.append('a')
    s.append('b')
    s.append('c')
    # stack is LIFO, thus the content of s is currently like this:
    # c <- top
    # b
    # a
    print("{:s}".format(s[-1]))                  # output 'c'
    s.pop();                                     # pop topmost
    print("{:s}".format(s[-1]))                  # output 'b'
    print("{:d}".format(len(s) == 0))            # s is not empty, false
    print("==================")

    q = deque([])
    print("{:d}".format(len(q) == 0))            # q is empty, true (1)
    print("==================")
    while len(s) != 0:                           # s has 2 more items
        q.append(s[-1]);                         # enqueue 'b' and 'a'
        s.pop();
    q.append('z')                                # add one more item
    print("{:s}".format(q[0]))                   # prints 'b'
    print("{:s}".format(q[-1]))                  # prints 'z'

    # output 'b', 'a', then 'z' (until queue is empty), according to the insertion order above
    print("==================")
    while len(q) != 0: 
        print("{:s}".format(q[0]))               # take the front first
        q.popleft();                             # before popping it


    print("==================")
    d = deque([])
    d.append('a')
    d.append('b')
    d.append('c')
    print("{:s} - {:s}".format(d[0], d[-1]))     # prints 'a - c'
    d.appendleft('d')
    print("{:s} - {:s}".format(d[0], d[-1]))     # prints 'd - c'
    d.pop()
    print("{:s} - {:s}".format(d[0], d[-1]))     # prints 'd - b'
    d.popleft()
    print("{:s} - {:s}".format(d[0], d[-1]))     # prints 'a - b'

if __name__ == "__main__":
    main()

File: ch2/lineards/list.java
import java.util.*;

class list {
  public static void main(String[] args) {
    Stack<Character> s = new Stack<>();
    System.out.println(s.isEmpty());             // s is empty, true
    System.out.println("==================");
    s.push('a');
    s.push('b');
    s.push('c');
    // stack is LIFO, thus the content of s is currently like this:
    // c <- top
    // b
    // a
    System.out.println(s.peek());                // output 'c'
    s.pop();                                     // pop topmost
    System.out.println(s.peek());                // output 'b'
    System.out.println(s.empty());               // s is not empty, false
    System.out.println("==================");

    // Queue is abstract, must be instantiated with LinkedList
    // (special case for Java Queue)
    Queue<Character> q = new LinkedList<>();
    System.out.println(q.isEmpty());             // q is empty, true
    System.out.println("==================");
    while (!s.isEmpty()) {                       // s  has 2 more items
      q.offer(s.peek()); // enqueue 'b', and then 'a' (the method name in Java Queue for push/enqueue operation is 'offer')
      s.pop();
    }
    q.offer('z');                                // add one more item
    System.out.println(q.peek());                // prints 'b'
    // in Java, it is harder to see the back of the queue...

    // output 'b', 'a', then 'z' (until queue is empty), according to the insertion order above
    System.out.println("==================");
    while (!q.isEmpty()) {
      System.out.printf("%c\n", q.peek());       // take the front first
      q.poll();                                  // before popping it
    }

    System.out.println("==================");
    Deque<Character> d = new LinkedList<>();
    d.addLast('a');
    d.addLast('b');
    d.addLast('c');
    System.out.printf("%c - %c\n", d.getFirst(), d.getLast()); // prints 'a - c'
    d.addFirst('d');
    System.out.printf("%c - %c\n", d.getFirst(), d.getLast()); // prints 'd - c'
    d.pollLast();
    System.out.printf("%c - %c\n", d.getFirst(), d.getLast()); // prints 'd - b'
    d.pollFirst();
    System.out.printf("%c - %c\n", d.getFirst(), d.getLast()); // prints 'a - b'
  }
}
